### DISCUSSION TITLE
In flight selection of waypoint mission

### DISCUSSION THREAD

**Discussion by tonyyng** [iNavFlight/inav#8360] [Ideas]
Hey @breadoven 

As I understand the multiple mission support, the mission can only be selected while on the ground.  Is that correct?

What if logic was added so it could be changed through inflight adjustments?  Besides the "I forgot to select the right mission" reason, I think it could be useful other ways.

What if your missions were setup as various landing approaches  to your flying fields, but from different directions.  Then you could use adjustments to select the one appropriate for the current wind direction.  It would be nice to be able to give a name to each mission as well.  "LAND EAST", or "6K Loop" and see that name displayed.

Thanks

**Comment by breadoven** [comment]
Currently you can only select missions on the ground. However, I did a PR recently that allows you to change missions in flight https://github.com/iNavFlight/inav/pull/8354. Uses the WP mode switch to cycle through the missions when a new "mission select" mode is active. I assign the mode to be active when Poshold is selected but you could use a separate switch if you wanted.

Named missions would be nice but doesn't seem so easy to do without rehashing the mission storage format.

**Reply by synersignart** [reply]
how is this setup , when you say cycle I only have a 2 point switch . ? also when I set the Mission select mode on and Flip the WP mode on and off I cant seem to see any indication the mission index change both in configurator mission tab or CLI , can someone pls clarify how this works

**Reply by breadoven** [reply]
It only works in the OSD. You need to have the `Mission Info` OSD field active which displays info on the selected mission number. No integration with Configurator as it stands. Need to update documentation on this ... still pending.

**Reply by synersignart** [reply]
any serial OSD recomendation . oddly I only have one here that works on mavlink

**Comment by tonyyng** [comment]
It might be more obvious changing the selected mission using the adjustments framework, but your solution will work.

Thanks

**Comment by tonyyng** [comment]
@breadoven I'm looking in your changes in #8354 and trying to understand how it works.   What logic causes the currently selected mission index to change?  

I see '''void selectMultiMissionIndex(int8_t increment)''', which is used by the stick commands.  I couldn't see how the mission select mode changes the mission index.

Thanks

**Comment by tonyyng** [comment]
nvm.  I see it in updateWpMissionChange()

**Comment by breadoven** [comment]
With #8354 it should be possible to just add an adjustment for the `nav_wp_multi_mission_index `setting. This should load the selected mission index when you're armed if it's different to the currently loaded mission index. Only minor drawback is it will load every time you change the index (unless you add something to only allow it to load when adjustments are finished).

This won't work for the current logic though because the mission waypoint array is overwritten when you arm with the selected mission moved to the start of the array so it always starts at waypoint index 0. You have to reload the mission when disarmed to refresh it. This is so all multi missions are available in Configurator when disarmed. #8354 avoids this overwriting by keeping the mission array intact and picking out the required mission based on start waypoint index for the selected mission instead which is a better way of doing it. When you disarm the start waypoint index is set to 0 and the number of waypoints set to the total in all multi missions making all missions available in Configurator.

**Reply by tonyyng** [reply]
>Only minor drawback is it will load every time you change the index (unless you add something to only allow it to load when adjustments are finished).
>

What is the drawback of doing this?  What if the adjustment to the mission_index was only allowed while NOT in wp mode?

I've found that with DJI HD, the options for changing mission index are more restricted.  You don't have access to the CMS menu.   With your missionindex change mode, you can't see the mission index because there isn't an OSD element for if. With adjustments at least you can display the mission index as you are changing it using the craftname hack.

**Reply by breadoven** [reply]
Drawback is just unwanted processing nothing more, not really an issue. You would need to inhibit WP mode I'd have thought whilst adjusting the mission index otherwise strange things might happen.

Is it not possible to add the Mission Info field to the DJI OSD, or something similar at least, perhaps just showing the currently selected mission index ? Not sure what the limitations are as to what can be displayed.

**Reply by tonyyng** [reply]
There are only certain fields that map directly to DJI OSD elements.  Others have to be sent in the craftname hack.  Those sent in the hack are displayed in a rotation, so only important fields are displayed through the hack.  Adjustments is one of those fields that is deemed important enough to display, while adjusting.

**Reply by breadoven** [reply]
So given it seems it's not easy to use multi missions with DJI stuff I'll add the mission index to adjustments. Looks like it's possible to know when the adjustment is active so perhaps it's possible to use this in a similar way to the Mission Change mode in #8354 to control things when changing the index.

**Comment by breadoven** [comment]
@tonyyng Can you try #8354 to see if the mission index adjustment works with DJI ? Needs Configurator support for the new Adjustment but you should be able enter it via CLI if you know the parameters (adjustment index is 59).

https://github.com/iNavFlight/inav/actions/runs/2987946351 is the last run.

**Comment by tonyyng** [comment]
My DJI enabled aircraft has a burned out motor.  I'm awaiting replacement parts before I can fly again.

What do I need in order to test it?  I'm currently on 4.1 (with customizations).  I had tried to manually merge #8354, before the latest changes into my repo, but it got too complicated.  There were too many changes since I branched off 4.1.

**Comment by breadoven** [comment]
You can use the target firmware from the inav-6.0.0-ci-20220904-11bc996.zip file in the link above and 6.0 Configurator from http://seyrsnys.myzen.co.uk/inav-configurator-next/.

The Adjustment won't appear in the Configurator Adjustment tab since it hasn't been added yet so easiest to set up some other Adjustment function first then just change the Adjustment function ID to 59 using the CLI, e.g. `adjrange 0 0 7 1600 1800 58 3` works for `FW_LEVEL_TRIM`, change the 58 to 59 to get it to work for Mission Index instead.

I assume when the Adjustment is active it appears in the DJI OSD field ? Should appear as WPI. If you can get the WPI number to change then it should load the selected mission when the Adjustment isn't active. Check `nav_wp_multi_mission_index `setting in the CLI to confirm the change.

**Comment by tonyyng** [comment]
I'm getting an error when I try to download Configurator 6.0:
Not Found
The requested URL /inav-configurator-next/files/INAV-Configurator_win64_6.0.0.zip was not found on this server.

**Reply by stronnag** [reply]
give me a minute to fix it. The configurator build has changed the file names

**Reply by stronnag** [reply]
should be fixed now

**Reply by tonyyng** [reply]
Thanks.

**Comment by tonyyng** [comment]
@breadoven  The adjustment selection of the mission index works as expected.  With 4 missions loaded, it would let me change the value in the range from 1 to 4.  In the DJI craftname field, I see WPI N as one of the messages when adjusting.

Thanks

**Comment by tonyyng** [comment]
@breadoven I see a hole in the checking of distance to the first waypoint. https://github.com/iNavFlight/inav/wiki/6.0.0-Release-Notes#increase-nav_wp_safe_distance-maximum

It is done against the current mission prior to arming, correct?  

What happens if you choose a different mission using the mode switch in-flight?  Is the check reapplied at that point?

I never did like the check prior to arming.  I think the check should be done when trying to enter wp mode.

**Reply by breadoven** [reply]
> @breadoven I see a hole in the checking of distance to the first waypoint. https://github.com/iNavFlight/inav/wiki/6.0.0-Release-Notes#increase-nav_wp_safe_distance-maximum
> 
> It is done against the current mission prior to arming, correct?
> 
> What happens if you choose a different mission using the mode switch in-flight? Is the check reapplied at that point?
> 
> I never did like the check prior to arming. I think the check should be done when trying to enter wp mode.

The distance check will only be at arming for the currently loaded mission. Won't work if you change the mission in flight. I don't really see this as much of issue if you have the System message showing in the OSD as you will immediately see the distance to the first waypoint when you start the mission and quickly realise something's wrong if it's too far away. Is this check even useful any more ? Back in the day distance to the next waypoint wasn't shown during a mission so you had no idea how far away it was, you only had the distance warning prior to arming to rely on.

**Comment by tonyyng** [comment]
Yes, I agree with you on the distance check's relevance.  I've modded my version to allow the max distance to be 5km so that it only comes into play when I've loaded the wrong mission for the current field.
